### PR TITLE
Fixed RuntimeError on unclosed async generators on py3.6/asyncio

### PR DESCRIPTION
**Pull request by agronholm**

### CODE REVIEW

File: src/anyio/_backends/_asyncio.py [hunk 1]
```diff
@@ -126,6 +128,14 @@ def current_task(loop: Optional[asyncio.AbstractEventLoop] = None) -> Optional[a
 
         return asyncio.Task.current_task(loop)
 
+
+# Needed for the workaround for async generator shutdown on Python 3.6
+async def _dummy_asyncgen():
+    yield
+
+async_generator_athrow = _dummy_asyncgen().athrow(Exception).__class__
```
Comment (graingert) [inline src/anyio/_backends/_asyncio.py:136]: Doesn't this throw an unawaited Coroutine warning?

File: src/anyio/_backends/_asyncio.py [hunk 2]
```diff
@@ -1857,7 +1875,9 @@ def _cancel_all_tasks(self) -> None:
             return
 
         for task in to_cancel:
-            task.cancel()
+            # Workaround for async generator shutdown on Python 3.6
+            if not isinstance(get_coro(task), async_generator_athrow):
+                task.cancel()
```
Comment (graingert) [inline src/anyio/_backends/_asyncio.py:1880]:
```suggestion
            # Workaround for async generator shutdown on Python 3.6
            if (
                sys.version_info >= (3, 6)
                and not isinstance(get_coro(task), async_generator_athrow)
            ):
                task.cancel()
```
you do want to cancel athrow tasks on 3.7+

File: src/anyio/_backends/_asyncio.py [hunk 3]
```diff
@@ -65,7 +66,9 @@ def _cancel_all_tasks(loop):
                 return
 
             for task in to_cancel:
-                task.cancel()
+                # Workaround for async generator shutdown on Python 3.6
+                if not isinstance(get_coro(task), async_generator_athrow):
+                    task.cancel()
```
Comment (graingert) [inline src/anyio/_backends/_asyncio.py:71]: you can `sleep(0)` and then cancel these tasks

File: src/anyio/_backends/_asyncio.py (hunk 1 above)
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py]: Fixed that one.

File: src/anyio/_backends/_asyncio.py (hunk 3 above)
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py:71]: How does `sleep(0)` help?

Comment (graingert) [reply src/anyio/_backends/_asyncio.py:71]: the task will have called `self._coro.send(None)` after sleep(0) finishes

**Comment by graingert** [comment]
Also the issue title is slightly wrong: the problem also applies to correctly aclosed async generators

File: src/anyio/_backends/_asyncio.py [hunk 4]
```diff
@@ -65,7 +66,9 @@ def _cancel_all_tasks(loop):
                 return
 
             for task in to_cancel:
-                task.cancel()
+                # Workaround for async generator shutdown on Python 3.6
```
Comment (graingert) [inline src/anyio/_backends/_asyncio.py:69]:
```suggestion
                # Workaround for async generator shutdown on Python 3.6
                # https://bugs.python.org/issue38013
```

File: src/anyio/_backends/_asyncio.py (hunk 2 above)
Comment (agronholm) [reply src/anyio/_backends/_asyncio.py:1880]: If we want to do that on 3.7.5+, wouldn't we be missing such tasks on earlier versions with the proposed code?

Comment (graingert) [reply src/anyio/_backends/_asyncio.py:1880]: missing?

Comment (agronholm) [reply src/anyio/_backends/_asyncio.py:1881]: Not cancelling them even though we possibly should.

Comment (graingert) [reply src/anyio/_backends/_asyncio.py:1881]: yes you have to wait an event loop turn to catch those missing tasks on 3.7.4-

**Comment by agronholm** [comment]
Given the rapidly approaching EOL of Python 3.6, this is likely not a fix we will incorporate into AnyIO.

**Pull request closed without merging**